internal/encryption/keyprovider/keyprovider.go

```diff
@@ -0,0 +1,22 @@
+package keyprovider
+
+type Config interface {
+	Build() (KeyProvider, error)
+}
+
+type Descriptor interface {
+	// ID returns the unique identifier used when parsing HCL or JSON configs.
+	ID() ID
+	// ConfigStruct creates a new configuration struct pointer annotated with hcl tags. The Build() receiver on
+	// this struct must be able to build a KeyProvider from the configuration:
+	//
+	// Common errors:
+	// - Returning a struct without a pointer
+	// - Returning a non-struct
+	ConfigStruct() Config
+}
+
+type KeyProvider interface {
+	// Provide provides an encryption key. If the process fails, it returns an error.
+	Provide() ([]byte, error)
```

Comment:
During encryption, key providers may need to place information in the "encryption" JSON key, such as a nonce used to derive the key.

During decryption, key providers may need to read information from the "encryption" JSON key, such as a nonce needed to derive the same key again.

Compare the KeyProvider interface in https://github.com/opentofu/opentofu/pull/1132/files#diff-e7749ff1b789f4930684fd4132a65fe3227b0a2dcbb23fe80794ddd5066cd88b this is why the parameters include an EncryptionInfo block, which later gets encoded into the "encryption" JSON substructure.

Reply:
Moreover, I believe this metadata should then be passed as "authenticated data" to the GCM Seal method.

Reply:
We now handle this by passing through metadata between keyproviders.